Fix type exports, run jobs on all pr's

BowlingX opened the pull request:
This will

- Bump `yarn` to latest
- Bump `zod` and `typescript` to latest
- Bumped `eslint` to `9.x`
- Adjust some scripts and escapings
- Adjust's the incorrect type exports (that where `*` exported in #16 )
- Make sure to always run tests on PR

BowlingX commented:
Hey @lostfictions, I applied some "project hygiene" and bumped dependencies.
I also adjusted that the builds run one each PR, otherwise we cannot see if the builds passed before merging.

I therefore had to adjust a change from a previously merged PR.

-----

I'm actually wondering regarding the release process.

I think we should automate that as well (e.g. when merging to `master`).
I usually use something like `semantic-release` to get propper changelogs etc., but we can also do that manually or semi automated (e.g. with release-it etc.)

lostfictions commented on package.json:
```diff
@@ -90,5 +92,6 @@
     }
-  }
+  },
+  "packageManager": "yarn@4.7.0"
```
would you have any objection if we used pnpm instead? it seems to have more momentum these days, and less quirks or caveats for contributors. we can use `pnpm import` to set up the lockfile.

i know it looks a little awkward in the `package.json` but we should also set the `packageManager` field via `corepack use`, which adds the sha512 sum (unfortunately corepack doesn't otherwise have a proper pinning/integrity mechanism besides this).

BowlingX replied:
Sure, we can do that :)

lostfictions commented on package.json:
```diff
@@ -39,38 +39,40 @@
-    "eslint-config-lostfictions": "^6.0.0",
+    "@types/jest": "^29.5.14",
+    "@types/node": "^22.13.10",
+    "@typescript-eslint/utils": "^8.27.0",
+    "eslint": "^9.22.0",
+    "eslint-config-lostfictions": "^7.0.0-beta.0",
+    "eslint-import-resolver-typescript": "^4.2.2",
```
`@typescript-eslint/utils` and `eslint-import-resolver-typescript` shouldn't be necessary, with the flat config format eslint configs can bring in their own dependencies

BowlingX replied:
Yeah, I had to add them, because yarn would complain otherwise, because they are implicitly pulled in by other debs, if we switch to pnpm that might not be a problem.

lostfictions commented on .github/workflows/build-and-test.yml:
```diff
@@ -13,15 +12,17 @@ jobs:
+    - run: yarn install --immutable
     - run: yarn build
     - run: yarn test
+    - run: yarn dlx @arethetypeswrong/cli . --pack
```
this should either be pinned to a specific sha or we should just add it as a `devDependency` imo

lostfictions commented on .github/workflows/build-and-test.yml:
```diff
@@ -13,15 +12,17 @@ jobs:
+    - uses: actions/checkout@v4
+    - run: corepack enable
     - name: Use Node.js ${{ matrix.node-version }}
-      uses: actions/setup-node@v3
+      uses: actions/setup-node@v4
       with:
         node-version: ${{ matrix.node-version }}
         cache: 'yarn'
-    - run: yarn install --frozen-lockfile
+    - run: corepack enable
```
`corepack enable` probably doesn't need to be invoked twice?

lostfictions commented on src/preprocessors.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-unsafe-argument,@typescript-eslint/no-unsafe-assignment */
```
i'll just disable the `no-unsafe-*` family in the eslint config and bump the version, they've ended up being a lot of friction for very little value :sweat_smile:

lostfictions commented on tsconfig.json:
```diff
@@ -1,11 +1,11 @@
+      "ES2022"
     ],
     "module": "Node16",
-    "moduleResolution": "Node16",
+    "moduleResolution": "node16",
```
i think we should maybe be switching this to at least `node18` now

BowlingX replied:
I believe there is only node16 (see, https://www.typescriptlang.org/tsconfig/#moduleResolution), however `module` can be `node18` from ts 5.8 https://www.typescriptlang.org/tsconfig/#node16node18nodenext, so I will use that.

lostfictions reviewed (commented):
thanks a lot for this! left a few comments, curious what you think

lostfictions commented:
> I'm actually wondering regarding the release process.
> 
> I think we should automate that as well (e.g. when merging to `master`). I usually use something like `semantic-release` to get propper changelogs etc., but we can also do that manually or semi automated (e.g. with release-it etc.)

No objections here to using automated release tooling like release-it. I do have a preference for an option that permits manually adding details or a top-level description to a changelog (rather than just pushing a list of merge commits, though I think it's good to automatically add the list of commits as "extra details", if that makes sense). I'll defer to you for however you'd like to set things up, in any event!

BowlingX commented:
> The only other request I have is that we merge #15 before merging this PR -- want to give it a quick review and merge if it seems ok to you? It's already been available as a prerelease on NPM for a while and it seems like at least a few people have tested that it works.

Yeah great, I will have a look, thank you!!

BowlingX commented:
I added another PR https://github.com/lostfictions/znv/pull/21, that will include your changes from #15. I had a look and I think it's alright. Regarding the type error you had, I adjusted the generic actually in my branch already (because I also had issues after updating zod). All checks still work properly.

lostfictions commented on eslint.config.mjs:
```diff
@@ -0,0 +1,13 @@
+import { defineConfig } from "eslint/config";
+
+import { default as lostificationConfig } from "eslint-config-lostfictions";
+
+export default defineConfig([
+  lostificationConfig,
+  {
+    rules: {
+      "@typescript-eslint/no-unsafe-argument": "off",
+      "@typescript-eslint/no-unsafe-assignment": "off",
+    },
+  },
+]);
```
i've just released `eslint-config-lostfictions@7.0.0` which disables them by default, so this should be fine again if you bump the package! Sorry for not explaining this more clearly.

```suggestion
export { default } from "eslint-config-lostfictions";
```

lostfictions commented:
Ok, merged `master` (including #15) in. Looks like there's still a type error in the test -- would you be up for taking a look?

BowlingX commented:
Thank you @lostfictions, I updated the types :).

lostfictions reviewed (approved):
looks great to me, thanks for all your work on this!

BowlingX commented:
I guess we can merge it then? I usually do squash merges to not pollute the git history too much, what do you prefer? And maybe you could release a new version then? Thank you!!

lostfictions commented:
Yep, I think you should have merge permissions so go ahead! Squash merges are good by me.

Agree that we should cut a new version, if you're interested in making a PR to add automated infra for that we can do that, otherwise I should be able to do it manually later today.

BowlingX commented:
@lostfictions, thank you, I would say we can do a manual release for now. I will then work on a new PR for the release automation.

lostfictions commented:
Done! https://github.com/lostfictions/znv/releases/tag/v0.5.0
